Editorial for Carrots for Rabbits (#6277)
cpinitiative/usaco-guide: solutions/silver/cf-1428E.mdx

Inuititively, we guess that the more pieces we cut a carrot into, the more the total time gets smaller.
And if we manually work out the time for each number of cuts, we see that the decrease in total time gets smaller as well.

To approximately see why, assume that the cuts are perfectly even, allowing fractional lengths;
in that case, $f(l, t)\approx t(\frac lt)^2=\frac{l^2}t$.
Note that since the function is in the form of $g(x)=\frac ax$, $f$ decreases and does so slower as $t$ increases;
this means that when incrementing the number of cuts on a carrot, each incrementation will save time, but the saving will decrease as we go further.

This inspires a greedy solution;
we have set carrots to be cut into $1$ segment initially (uncut), and precompute the decrease in time of that carrot if the number of cuts on it is increased by $1$.
We maintain a priority queue of carrots sorted by the change described above;
for $k-n$ times, we take the carrot with the largest change value and increase its number of cuts by $1$.
We then recompute the time of eating that carrot, and the change to that time if the number of cuts is *further* increased by $1$ before reenqueing the carrot.

> As the returns are diminishing for each increase in the cuts, there cannot be a situation where picking a suboptimal cut first will result in better cuts later for a carrot.

Reviewer: ```suggestion
The sum of the squares of two numbers given a fixed sum is minimized when the two numbers are equal.
If we extend that idea to each carrot, then they should be cut into pieces of lengths that are as close together as possible, without any pair of lengths differing by more than $1$.
Specifically, if a carrot of length $l$ is cut into $t$ parts, there will be $t-l\bmod t$ pieces with length $\lfloor\frac lt\rfloor$, and $l\bmod t$ pieces with length $\lfloor\frac lt\rfloor+1$.
Let the time taken to eat these pieces be $f(l,p)$, which can be calculated in $\mathcal O(1)$ time.

Now, the problem reduces to deciding the number of cuts on each carrot.
Intuitively, the more pieces we cut a carrot into, the smaller the total time gets.
And if we manually calculate the time for each number of cuts, we see that the decrease in time after each cut gets smaller as well.

To see why, assume that the cuts are perfectly even, allowing fractional lengths.
In that case, $f(l, t)\approx t(\frac lt)^2=\frac{l^2}t$.
Since the function is in the form of $g(x)=\frac ax$, $f$ decreases and does so more slowly as $t$ increases.
This means that when incrementing the number of cuts on a carrot, each incrementation will save time, but the saving will decrease as we go further.

This inspires a greedy solution, where we set carrots to be cut into $1$ segment initially (uncut), and precompute the decrease in time of that carrot if the number of cuts on it is increased by $1$.
We maintain a priority queue of carrots sorted by the decrease in time described above;
for $k-n$ times, we take the carrot with the largest change value and increase its number of cuts by $1$.
We then recompute the time of eating that carrot, and the change to that time if the number of cuts is *further* increased by $1$ before placing the carrot back in the priority queue.

As the returns are diminishing for each increase in the cuts, there cannot be a situation where picking a suboptimal cut first will result in better cuts later for a carrot.
```

- some grammar/writing things
- im personally not a semicolon fan gg

Author: Thanks.  I am a semicolon fan as I've been brainwashed by my ELA teacher to believe that semicolons are extremely valuable for those "correct the sentences" exercises on standardized exams.

Author: (Brutal example of how the educational system, forced to tailor to standardized tests, fails to prepare students for life.

---
id: cf-1428E
source: CF
title: Carrots for Rabbits
author: John Zhou
---

[Official Editorial (C++, Python)](https://codeforces.com/blog/entry/83771)

<Spoiler title="Hint 1">

If you have a carrot and know the number of pieces you want to cut it into, how would you cut it to minimize the time to eat for all the segments of that carrot?

</Spoiler>

<Spoiler title="Hint 2">

> Try cutting a carrot into a number of different numbers of segments.  What do you observe about the total time to eat for that carrot?

Reviewer 2: ```suggestion
Try cutting a carrot into a number of different segments.  What do you observe about the total time to eat that carrot?
```

Author: @Sosuke23 Here I mean literally a variety of numbers of segements.  So I've corretced that to get rid of the duplicate "numbers" and used "variety".

```suggestion
Try cutting a carrot into a variety of numbers of segments.  What do you observe about the total time to eat that carrot?
```

</Spoiler>

<Spoiler title="Hint 3">

> In hint $2$, what do you observe about the *change* in the total time to eat for that carrot?

Reviewer 2: ```suggestion
In hint $2$, what do you observe about the *change* in the total time to eat that carrot?
```

</Spoiler>

> ## Explanation

Reviewer 2: Since this solution has hints, put the explanation under the spoiler tag too.

To minimize time, each carrot should be cut into pieces of lengths that are as close together as possible.
Specifically, no two pieces should differ by more than $1$ in length.
Thus, if a carrot of length $l$ is cut into $t$ parts, there will be $t-l\bmod t$ of pieces with length $\lfloor\frac lt\rfloor$, and $l\bmod t$ pieces with length $\lfloor\frac lt\rfloor+1$.
This allows us to calculate the optimal eating time for a carrot with a specific number of cuts in $\mathcal O(1)$ time;
let the result of this calculation be $f(l,p)$.

> <Spoiler title = "Proof: Optimal lengths differ by at most $1$">

Reviewer 2: after rendering, the title is weird. maybe keep it normal 

<img width="1125" height="179" alt="Image" src="https://github.com/user-attachments/assets/c8d4dafb-9f67-4824-865d-cac6fbb673c4" />

Author: Do you mean like the LaTeX formatting should be changed to normal?

Author: I've done that for now; would y'all appreciate a PR to render KaTeX in spoiler titles?

Author: Resolving this for now to unblock.

---
id: cf-1428E
source: CF
title: Carrots for Rabbits
author: John Zhou
---

[Official Editorial (C++, Python)](https://codeforces.com/blog/entry/83771)

<Spoiler title="Hint 1">

If you have a carrot and know the number of pieces you want to cut it into, how would you cut it to minimize the time to eat for all the segments of that carrot?

</Spoiler>

<Spoiler title="Hint 2">

> Try cutting a carrot into a variety of numbers of segments.  What do you observe about the total time to eat that carrot?

Reviewer: wording here could be improved

maybe like, "for a carrot of arbitrary length, calculate the cost of cutting it 1, 2, 3, and more times. what do you notice about the cost for each time?" or something like that

Author: Done

Specifically, if a carrot of length $l$ is cut into $t$ parts, there will be $t-l\bmod t$ of pieces with length $\lfloor\frac lt\rfloor$, and $l\bmod t$ pieces with length $\lfloor\frac lt\rfloor+1$.
Let the time taken to eat these pieces be $f(l,p)$, which can be calculated in $\mathcal O(1)$ time.

<Spoiler title = "Proof: Optimal lengths differ by at most 1">

Suppose two pieces differ by at least $2$.  One can use another cut with $1$ more length to the smaller piece, and $1$ less length to the larger piece;
this preserves the sum of the lengths, which also ensures that the other pieces' arrangements are not affected;
in addition, this reduces the difference between the lengths.

Let the old lengths of the carrots be $X$ and $Y$, and the new lengths of the carrots be $X'$ and $Y'$.
Since $X+Y=X'+Y'$ and $\lvert X-Y\rvert > \lvert X'-Y'\rvert$, we have

$$
\begin{aligned}
X^2 + Y^2 &= \frac{(X+Y)^2+(X-Y)^2}2 = \frac{(X'+Y')^2+(\lvert X-Y\rvert)^2}2 \\
          &> \frac{(X'+Y')^2+(\lvert X'-Y'\rvert)^2}2 = (X')^2+(Y')^2.
\end{aligned}
$$

> The eating time for the $2$ segments is thus reduced; since other segments are unaffected, the overall time is reduced as well.

Reviewer: is it not just like

$$
x^2+y^2 \geq (\lceil \frac{x+y}{2} \rceil)^2 + (\lfloor \frac{x+y}{2} \rfloor)^2
$$

i think that'd be a lot easier to understand than the current version

Author: I don't get how this equation works in the context of this problem?

Reviewer: we seek to prove that $|x - y| \leq 1$ is optimal for splitting

if $x$ and $y$ are the two carrot sizes we make from splitting, we show that by instead dividing them up like this (ie so their differences are = 0 or =1 by taking the average) that this way is at least as good or better than the current sol

Author: Well I wanted to show the math rather than just stating the conclusion.

Do you think dropping the math here is worthwhile? If so, how much?

Reviewer: no like the inequality is just one way of proving the conclusion that is fairly straightforward (which I think is easier to understand? again another opinion would be nice)

Author: @eysbutno I get it now... Note to self, this means that the sum of squares of the average of the two numbers rounded up and down which is the most it's ever going to be adjusted to.

But now the rounding gets a bit hard to explain rigorously.  So waiting on another opinion for whether this or the incremental appraoch is more intuitive.

Reviewer 3: i prefer john's approach here tbh but the way he wrote his proof is a lil unintuitive (imo) cuz like writing the sum of squares like that isnt the most intuitive ykwim i think if you frame it as something like 

If the total carrot size is S, the perfect average is A = S/2. Any split can be written as its distance (d) from that average: x = A + d and y = A - d. 

When you calculate the sum of squares, the middle terms cancel out:
x^2 + y^2 = (A + d)^2 + (A - d)^2 = 2A^2 + 2d^2

Since 2A^2 is a fixed value, minimizing the sum of squares completely depends on making 2d^2 as small as possible. 

Because we are working with integers, d can't always be 0:
- If S is even, the average is a whole number, so we choose d = 0. This gives |x - y| = 0.
- If S is odd, since x and y are integers, the closest we can get is a distance of d = 0.5. This gives x = A + 0.5 (the ceiling) and y = A - 0.5 (the floor), leaving a difference of |x - y| = 1.

Therefore, keeping |x - y| <= 1 mathematically guarantees the lowest possible sum of squares.

it'd be a bit more intuitive

Reviewer 3: But like it is a well-known property that minimizing the sum of squares (or maximizing the product) under a fixed sum requires keeping the elements as close to each other as possible, the optimal strategy is to make their differences <=1, so id probably not prove this in the contest

Reviewer 3: > i prefer john's approach here tbh but the way he wrote his proof is a lil unintuitive (imo) cuz like writing the sum of squares like that isnt the most intuitive ykwim i think if you frame it as something like
> 
> If the total carrot size is S, the perfect average is A = S/2. Any split can be written as its distance (d) from that average: x = A + d and y = A - d.
> 
> When you calculate the sum of squares, the middle terms cancel out: x^2 + y^2 = (A + d)^2 + (A - d)^2 = 2A^2 + 2d^2
> 
> Since 2A^2 is a fixed value, minimizing the sum of squares completely depends on making 2d^2 as small as possible.
> 
> Because we are working with integers, d can't always be 0:
> 
> * If S is even, the average is a whole number, so we choose d = 0. This gives |x - y| = 0.
> * If S is odd, since x and y are integers, the closest we can get is a distance of d = 0.5. This gives x = A + 0.5 (the ceiling) and y = A - 0.5 (the floor), leaving a difference of |x - y| = 1.
> 
> Therefore, keeping |x - y| <= 1 mathematically guarantees the lowest possible sum of squares.
> 
> it'd be a bit more intuitive

we can naturally extend this argument to prove that f(l,t) is minimized when each element is either floor(l/t) or ceil(l/t) and since no further restrictions the floor(l/t) must come l/t times, and ceil(l/t) only l%t times and like the proof can be shortened heavily imo

Author: Removed the entire block as intuition is better than proof and the rest of the content already explains the intuition.

</Spoiler>

> <Spoiler title = "Proof: Construction for f(l, p)">

Reviewer: what's the purpose of this section? I don't quite understand

Author: I proved that no optimal construction can have lengths more than 2 apart, but I wanted to demonstrate that there is only one way (the remainder method) to construct that goal in this section.

Reviewer: i don't think this is super necessary (would like to hear another opinion tho)

Author: Removed the entire block as intuition is better than proof and the rest of the content already explains the intuition.

[Euclidean Division Lemma](https://en.wikipedia.org/wiki/Euclidean_division#Division_theorem)
directly implies the aforementioned construction.

</Spoiler>

Now, the problem reduces to deciding the number of cuts on each carrot.
One may think that the overall set of piece lengths across all original carrots should be as close together as possible, just like we did for the piece lengths in a single carrot.
However, this is not generally doable, as one may have small segments that are below the ideal segment length that will ensure all cuts are equal.

We note that the more pieces we cut a carrot into, the more the total time gets smaller;
however, the returns of changing to cut one extra segment is diminishing as the number of already cut segments increase.
To see why, assume that the cuts are perfectly even, allowing fractional lengths;
in that case, $f(l, t)\approx t(\frac lt)^2=\frac{l^2}t$.
Note that since the function is in the form of $g(t)=\frac ct$ when $l$ is held constant, the function decreases slower as $l$ increases;
this means that when incrementing the number of cuts on a carrot, each incrementation will save time, but the saving will decrease as we go further.

> <Spoiler title = "Proof: More pieces equals less total time">

Reviewer: i think it'd be more logical to organize it like:
- more cuts is less time (ok i think this is intuitively obvious enough)
- each cut has less saving (this is less obvious, put proof in a spoiler)

Author: Done.  I have an intuitive way to reason about this and then formal proofs.  Neither of the formal proofs are trivial, but one can handwave and arrive at these conclusions fairly quickly.

Now, we use a similar logic on the left side; $f(l, p-1) + f(l, p+1)$ is a way to cut $2l$ into $2p$ as well, shown trivially if one simply glues the two carrots together.
This means that it'll be larger than or equal to the optimum, $f(2l, 2p)$, proving our proposition.

</Spoiler>

This inspires a greedy solution;
we have set carrots to be cut into $1$ segment initially (uncut), and precompute the decrease in time of that carrot if the number of cuts on it is increased by $1$.
We maintain a priority queue of carrots sorted by the change described above;
for $k-n$ times, we take the carrot with the largest change value and increase its number of cuts by $1$.
We then recompute the time of eating that carrot, and the change to that time if the number of cuts is *further* increased by $1$ before reenqueing the carrot.

This is guaranteed to be optimal;
as the returns are diminishing for each increase in the cuts, there cannot be a situation where picking a suboptimal cut first will result in better cuts later for a carrot.

## Implementation

**Time Complexity:** $\mathcal{O}(k\log n)$

<LanguageSection>
<CPPSection>

```cpp

> #include <iostream>

Reviewer: ok i feel like the way this code is written is kind of weird to read for a typical competitive programmer

https://codeforces.com/contest/1428/submission/379068850

here's my code, IMO smth like this with comments would probably be much more readable

Author: Says I'm not allowed to view the requested page.

Do you think I should uncurry updateCache (i.e. make it a global function instead of a class member?). Anything else

Reviewer: you have to log into cf to view submissions

here's the code

```cpp
#include <bits/stdc++.h>

using ll = long long;

void solve() {
	int n, k;
	std::cin >> n >> k;
	std::priority_queue<std::array<ll, 3>> pq;
	ll res = 0;

	auto cost = [&](int len, int cuts) -> ll {
		ll avg = len / (cuts + 1);
		int diff = len % (cuts + 1);
		return 1ll * avg * avg * (cuts + 1 - diff)
			 + 1ll * (avg + 1) * (avg + 1) * diff;
	};

	for (int i = 0; i < n; i++) {
		int a;
		std::cin >> a;
		res += 1ll * a * a;
		pq.push({cost(a, 0) - cost(a, 1), a, 1});
	}

	for (int i = 0; i < k - n; i++) {
		auto [diff, len, cuts] = pq.top();
		pq.pop();
		res -= diff;
		pq.push({cost(len, cuts) - cost(len, cuts + 1), len, cuts + 1});
	}

	std::cout << res << '\n';
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    solve();
}
```
